### ISSUE TITLE
Add "Add to cart" function via ajax in product list and product details

### ISSUE THREAD

**Issue by nrunde** [isotope/core#642]
I don't know if there are already plans to implement a feature like this.
But I think it is a important one for modern shop systems so isotope need to, imho.

**Comment by bytehead** [comment]
@aschempp any hints how to approach that? i would provide a PR as well, if i get this functionality to work...

**Comment by aschempp** [comment]
Difficult to say as I have not ever pursued this. The first problem is that list/reader are totally independent of the cart module. So clicking the cart button in one module would potentially need to update other modules on the same page. Then the cart rows can be customized in the template, which will make it hard to do client side.

Adding things to the cart in the ajax request is fairly simple, just check the button callbacks. But updating the frontend is probably a pain…

**Comment by Toflar** [comment]
In the FE I guess you cannot just reload e.g. the cart front end module because you also have other parts that depend on it like insert tags (e.g. `{{cart::items}}` which would need to be updated).

Personally, I would do it like this:

* FE modules that have buttons now (list, variant list, reader, related products) get a new checkbox "Execute button actions via ajax"
* On activation of that checkbox you get a "Update CSS selector" text field.

In this CSS selector you could for example enter `[data-isotopeajax="updatemenow"]` or something simpler like `div.css-class-update-me-on-ajax`.

Basically what the FE module then would do is request the whole page again and filter both, the ajax request DOM and the current page DOM for the CSS selector and replace all the contents accordingly. That way I could also make sure my InsertTags get updated by writing something like `<div class="css-class-update-me-on-ajax">{{cart::items}}</div>`.

This is the only way I can think of it being implemented with the current architecture. I just wonder if this is really worth the efforts because it would be implementing ajax just for the sake of "eye candy" as it doesn't speed up things at all because the request still loads all the html. In future, one day Contao 4 implements all the 4 ways to include content (fe modules, content elements, articles, intert tags) as separate controllers, we can start using ESI and it's a whole different story then.

**Comment by aschempp** [comment]
I didn't even think about the insert tags. What if they would output some HTML to account for an update?

**Comment by Toflar** [comment]
It would be updated automatically?

**Comment by bytehead** [comment]
Thanks for the hints :+1:!
I have now created a custom module which does handle my ajax requests.
You can choose the corresponding cart on the module within a select field (`$this->ajaxCart`). 

If I get a FMD via `Controller::getFrontendModule($intId)` the insert tags should work afaik.

It actually works with a few lines: 
````php

protected function compile()
    {
        if (\Environment::get('isAjaxRequest')) {
            if (\Input::post('isotope') == 'updateCart') {
                $fmd = \Controller::getFrontendModule($this->ajaxCart);

                $response = new JsonResponse([
                    'data' => $fmd,
                ]);

                $response->send(true);
            }

            if ((int) \Input::post('remove') > 0) {
                Isotope::getCart()->deleteItemById((int) \Input::post('remove'));
            }
        }
    }
````

**Comment by Toflar** [comment]
Yes, I know it is very simple if you develop it on your own. I've implemented several ajax solutions together with Isotope but never a general one that can be enabled by click and shipped with the Core because it's not that simple. But I keep saying this in the Community Forums: You should rather choose Isotope as a foundation for your own implementations instead of expecting it to be your out-of-the-box solution. It was never meant to be one.

**Comment by bytehead** [comment]
:+1: it's fine for me!
